pgjdbc/src/main/java/org/postgresql/core/v3/replication/V3PGReplicationStream.java

```diff
@@ -222,6 +222,9 @@ private void updateStatusInternal(
 
   private boolean processKeepAliveMessage(ByteBuffer buffer) {
     lastServerLSN = LogSequenceNumber.valueOf(buffer.getLong());
+    if (lastServerLSN.asLong() > lastReceiveLSN.asLong()) {
+      lastReceiveLSN = lastServerLSN;
+    }
```

Comment:
This mimics [the keepalive logic in pg_recvlogical](https://github.com/postgres/postgres/blob/REL_10_1/src/bin/pg_basebackup/pg_recvlogical.c#L473) where `output_written_lsn` is set to `Max(walEnd, output_written_lsn)` and `walEnd` is the first long of the server keepalive message.

It's unclear to me whether this logic also makes sense for physical replication, or if different logic is needed there.

Reply:
I seriously doubt anyone uses the physical replication facility in JDBC